lib/srv/db/common/packettrace/tracing.go

```diff
@@ -0,0 +1,235 @@
+	TeleportToServer
+)
+
+const fakeClientAddr = "1.1.1.1"
+const fakeTeleportAddr = "2.2.2.2"
+const fakeServerAddr = "3.3.3.3"
+
+const mergecapBin = "mergecap"
+const text2pcapBin = "text2pcap"
+
+func (p Direction) String() string {
+	switch p {
+	case ClientToTeleport:
+		return "Client->Teleport"
+	case ServerToTeleport:
+		return "Server->Teleport"
+	case TeleportToClient:
+		return "Teleport->Client"
+	case TeleportToServer:
+		return "Teleport->Server"
+	default:
+		return fmt.Sprintf("Unknown(%d)", p)
+	}
+}
+
+// PacketEntry holds the details of each packet to be recorded.
+type PacketEntry struct {
+	Direction Direction
+	Payload   []byte
+	Timestamp time.Time
+}
+
+// Trace struct holds all packet details and offers methods to add and save packets.
+type Trace struct {
+	packets    []PacketEntry
+	clock      clockwork.Clock
+	runCommand func(name string, arg ...string) ([]byte, error)
+
+	mx sync.Mutex
+}
+
+// NewTrace initializes a new Trace object.
+func NewTrace(clock clockwork.Clock) *Trace {
+	return &Trace{
+		packets: make([]PacketEntry, 0),
+		clock:   clock,
+		runCommand: func(name string, arg ...string) ([]byte, error) {
+			cmd := exec.Command(name, arg...)
+			out, err := cmd.CombinedOutput()
+			return out, trace.Wrap(err)
+		},
+	}
+}
+
+// AddPacket adds a packet to the trace based on the sender (client or server) and the payload.
+// It automatically handles source and destination addresses based on the sender.
+func (t *Trace) AddPacket(direction Direction, payload []byte) {
+	// Record timestamp
+	timestamp := t.clock.Now()
+
+	// Create a new PacketEntry
+	packet := PacketEntry{
+		Direction: direction,
+		Payload:   payload,
+		Timestamp: timestamp,
+	}
+
+	// Add packet to the trace
+	t.mx.Lock()
+	t.packets = append(t.packets, packet)
+	t.mx.Unlock()
+}
+
+type participant struct {
+	addr      string
+	direction Direction
+}
+
+func (t *Trace) saveOneLinkToPCAP(filename string, port int, sender, receiver participant, packets []PacketEntry) error {
+	buffer := &bytes.Buffer{}
+	for _, packet := range packets {
+		// assign indicator or skip packet.
+		indicator := ""
+		if packet.Direction == sender.direction {
+			indicator = "I"
+		}
+		if packet.Direction == receiver.direction {
+			indicator = "O"
+		}
+		if indicator == "" {
+			continue
+		}
+
+		// Write the timestamp and sender indicator
+		if _, err := fmt.Fprintf(buffer, "%s %s\n", indicator, packet.Timestamp.UTC().Format(time.RFC3339Nano)); err != nil {
+			return trace.Wrap(err)
+		}
+
+		// Write the packet data using hex.Dump and add a newline after each packet
+		hexData := hex.Dump(packet.Payload)
+		if _, err := buffer.Write([]byte(hexData + "\n")); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	filenameHex := filename + ".hex"
+
+	err := os.WriteFile(filenameHex, buffer.Bytes(), 0600)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer os.Remove(filenameHex)
+
+	// Invoke text2pcap to convert the hex dump to PCAP format
```

Comment:
I assume this isn't used for anything where performance matters, right?

Writing in one format and using an external utility to get to the desired format seems rather inefficient, but I don't have any context on what this is for.

Reply:
Nope, this is meant for debugging. I'll add comments about this.

Reply:
I've added some disclaimers. 

In practice I don't think this would really be that slow, text2pcap is an efficient C program using `libpcap` under the hood, it may well be faster than equivalent Go code even with the intermediate step. Nonetheless, it shouldn't be used in performance critical contexts, and I've added respective comments.